File: inkwire/receivers/python-fastapi/inkwire_receiver/store.py
from typing import Protocol, Optional, TypedDict
class StoredPost(TypedDict):
    id: str; slug: str; url: str; external_id: str
class UpsertResult(TypedDict):
    id: str; url: str; slug: str; created: bool
class PostStore(Protocol):
    async def find_by_external_id(self, key: str, external_id: str) -> Optional[StoredPost]: ...
    async def find_by_slug(self, key: str, slug: str) -> Optional[StoredPost]: ...
    async def upsert(self, key: str, post: dict) -> UpsertResult: ...
class MemoryStore:
    def __init__(self): self._by_ext = {}; self._by_slug = {}; self._seq = 0
    def _k(self, key, i): return f"{key}::{i}"
    async def find_by_external_id(self, key, external_id):
        return self._by_ext.get(self._k(key, external_id))
    async def find_by_slug(self, key, slug):
        return self._by_slug.get(self._k(key, slug))
    async def upsert(self, key, post):
        existing = self._by_ext.get(self._k(key, post["external_id"]))
        if existing: pid = existing["id"]
        else: self._seq += 1; pid = str(self._seq)
        slug = post["slug"]; url = f"https://demo.local/blog/{slug}"
        rec = {"id": pid, "slug": slug, "url": url, "external_id": post["external_id"]}
        self._by_ext[self._k(key, post["external_id"])] = rec
        self._by_slug[self._k(key, slug)] = rec
        return {"id": pid, "url": url, "slug": slug, "created": existing is None}

File: inkwire/receivers/core-ts/src/store.ts
export interface StoredPost { id: string; slug: string; url: string; external_id: string; }
export interface UpsertInput {
  external_id: string; title: string; html: string; markdown: string; slug: string;
  excerpt?: string; tags?: string[]; cover_image_url?: string; canonical_url?: string;
  author?: { name?: string; email?: string }; status: "draft" | "published"; published_at: string;
}
export interface PostStore {
  findByExternalId(key: string, externalId: string): Promise<StoredPost | null>;
  findBySlug(key: string, slug: string): Promise<StoredPost | null>;
  upsert(key: string, post: UpsertInput): Promise<{ id: string; url: string; slug: string; created: boolean }>;
}
export class MemoryStore implements PostStore {
  private byExt = new Map<string, StoredPost>();
  private bySlug = new Map<string, StoredPost>();
  private seq = 0;
  private k(key: string, id: string) { return `${key}::${id}`; }
  async findByExternalId(key: string, externalId: string) { return this.byExt.get(this.k(key, externalId)) ?? null; }
  async findBySlug(key: string, slug: string) { return this.bySlug.get(this.k(key, slug)) ?? null; }
  async upsert(key: string, post: UpsertInput) {
    const existing = this.byExt.get(this.k(key, post.external_id));
    const id = existing?.id ?? String(++this.seq);
    const slug = post.slug;
    const url = `https://demo.local/blog/${slug}`;
    const rec: StoredPost = { id, slug, url, external_id: post.external_id };
    this.byExt.set(this.k(key, post.external_id), rec);
    this.bySlug.set(this.k(key, slug), rec);
    return { id, url, slug, created: !existing };
  }
}
